Case search pillow

proteusvacuum opened the pull request:
Starting point for case search. 
Leaving the do not merge on as want to get feedback. 
This also includes a janky debug view to test searches (will work on the UI for this later, this was just to get something out there)

@kaapstorm - haven't added your feedback from the last branch yet 
@czue - FYI

cc: @gcapalbo @orangejenny 

(@esoergel could you check over the ES stuff -- there will be more additions when I add the ability to boost different properties, but this is a starting point).

proteusvacuum committed:
Add case search pillow

proteusvacuum committed:
Add app to track Case search configs

proteusvacuum committed:
Add nested queries to ESQuery

proteusvacuum committed:
Enable reindexing with v2 reindexer

proteusvacuum committed:
Add case search pillows and app to settings

proteusvacuum committed:
CaseSearchES

proteusvacuum committed:
Add case search to ES_META

proteusvacuum committed:
Optimize imports

proteusvacuum committed:
Add tests

proteusvacuum committed:
Add debug view

proteusvacuum committed:
Delete cases from index when disabling case search

proteusvacuum committed:
Add pillow to all-pillow-meta

dimagimon commented on corehq/pillows/case_search.py:
```diff
@@ -0,0 +1,149 @@
+        domain = self._get_domain(change)
+        change_object = self._get_change_object(change)
+        if domain and case_search_enabled_for_domain(domain):
+            super(CaseSearchPillowProcessor, self).process_change(pillow_instance, change_object, do_set_checkpoint)
```
E501 line too long (116 > 115 characters)

dimagimon commented on testapps/test_pillowtop/tests/test_case_search_pillow.py:
```diff
@@ -0,0 +1,139 @@
+        ensure_index_deleted(CASE_SEARCH_INDEX)
+
+        # Bootstrap ES
+        case_pillow = CaseSearchPillow()
```
F841 local variable 'case_pillow' is assigned to but never used

dimagimon commented on testapps/test_pillowtop/tests/test_case_search_pillow.py:
```diff
@@ -0,0 +1,139 @@
+
+        # enable case search for domain
+        with patch('corehq.pillows.case_search.case_search_enabled_for_domain',
+                        new=MagicMock(return_value=True)) as fake_case_search_enabled_for_domain:
```
E127 continuation line over-indented for visual indent

dimagimon commented on testapps/test_pillowtop/tests/test_case_search_pillow.py:
```diff
@@ -0,0 +1,139 @@
+        CaseSearchConfig.objects.get_or_create(pk=other_domain, enabled=True)
+
+        desired_case = self._make_case(domain=other_domain)
+        undesired_case = self._make_case(domain=self.domain)
```
F841 local variable 'undesired_case' is assigned to but never used

proteusvacuum committed:
Add case_search to expected prod indices

proteusvacuum committed:
lint

proteusvacuum committed:
Add fuzzy matching

proteusvacuum committed:
Make MUST the default clause

proteusvacuum committed:
Add fuzzy matching to UI with some cleanup

proteusvacuum committed:
Add fuzzy properties

proteusvacuum committed:
Add test and fix typo

dimagimon commented:
:+1: No lint errors found.

proteusvacuum committed:
Call reindex properly

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into uatbc/case-search-es

proteusvacuum committed:
Move api to models

czue commented on corehq/apps/case_search/models.py:
```diff
@@ -0,0 +1,118 @@
+        primary_key=True
+    )
+    enabled = models.BooleanField(blank=False, null=False, default=False)
+    config = JSONField(default={})
```
I see we do this everywhere in the code but I think it's bad practice. The default should be `dict` so that everything doesn't point to the same instance. https://docs.djangoproject.com/en/1.9/ref/models/fields/#default

czue commented on corehq/apps/case_search/models.py:
```diff
@@ -0,0 +1,118 @@
+                }
+            ]
+        }
+        """
```
I think it'd be nice if `self.config` was a `JsonObject`. Then we could clearly define the schema we're expecting and push operations on the config into the config object itself. You could trivially do this by just adding a `wrapped_config` property to the model, or (out of scope) you could do something fancy like make a `JsonObjectField` subclass of `JsonField`.

czue commented on corehq/apps/case_search/models.py:
```diff
@@ -0,0 +1,118 @@
+
+        raise AttributeError("{} is not a fuzzy property for {}".format(property, case_type))
+
+    def fuzzy_properties_for_case_type(self, case_type):
```
nit: might be clearer if this starts with `get_`

czue commented on corehq/apps/case_search/templates/case_search/case_search.html:
```diff
@@ -0,0 +1,186 @@
+{% endcompress %}
+{% block js-inline %}
+    <script>
+     var CaseSearchModel = function(){
```
@dimagi/js-team is something of this size recommended to be moved to an external JS file?

orangejenny replied:
would be nice, yeah

czue commented on corehq/apps/case_search/templates/case_search/case_search.html:
```diff
@@ -0,0 +1,186 @@
+{% endblock %}
+{% block page_content %}
+    <div id="case-search">
+        <h3>Search Parameters</h3>
```
file is missing translations

czue commented on corehq/apps/case_search/views.py:
```diff
@@ -0,0 +1,28 @@
+    template_name = 'case_search/case_search.html'
+    urlname = 'case_search'
+
+    def get(self, request, *args, **kwargs):
```
I think this needs some permission controls

czue commented on corehq/pillows/case_search.py:
```diff
@@ -0,0 +1,151 @@
+    es_index = CASE_SEARCH_INDEX
+    default_mapping = CASE_SEARCH_MAPPING
+
+    def change_trigger(self, changes_dict):
```
if the goal is for this pillow to be a noop then I think it's better to override `run()` so that it doesn't even have to stream through any changes

czue commented on corehq/pillows/case_search.py:
```diff
@@ -0,0 +1,151 @@
+    return doc
+
+
+def _get_case_properties(doc_dict):
```
@snopoke is there a standard way to do this (get all case dynamic properties) that is couch and sql compatible?

czue replied:
actually it looks like `.dynamic_case_properties` is what should be used.

snopoke replied:
yes, `.dynamic_case_properties()` is the correct method

proteusvacuum replied:
@snopoke how do I do this? Does `CommCareCase.wrap(doc_dict).dynamic_case_properties()` make sense for SQL too? Not sure if `wrap` is meaningful in the SQL context, or even if the pillows receive a dict in the SQL case?

czue replied:
interesting, i'm not sure there is an easy way to go from the dict to a properly wrapped model. it would be nice if there was a standardized way to do this from the dicts themselves. maybe we can start that here and have a todo that it needs to be updated to deal with SQL better?

benrudolph replied:
we should actually be able to just use the django rest framework to deserialize the `doc_dict`. might take a little massaging, but i think that'd be the proper way to do it given we use it for serialization: https://github.com/dimagi/commcare-hq/blob/master/corehq/form_processor/serializers.py

some docs on deserialization: http://www.django-rest-framework.org/api-guide/serializers/

proteusvacuum replied:
I think I can just go with something like: 
```py
if should_use_sql_backend(domain):
        doc_dict.pop("_id")
        dynamic_case_properties = CommCareCaseSQL(**doc_dict).dynamic_case_properties()
else:
        dynamic_case_properties = CommCareCase.wrap(doc_dict).dynamic_case_properties()
```

proteusvacuum replied:
I should probably abstract that out somewhere...

snopoke replied:
I don't think that will work for SQL but just looking at the json serialization I think you could maybe just go `dynamic_case_properties = doc_dict['case_json']`

See `CommCareCaseSQL.to_json`

proteusvacuum replied:
:+1:

czue commented on corehq/pillows/case_search.py:
```diff
@@ -0,0 +1,151 @@
+                pillow_instance, change_object, do_set_checkpoint
+            )
+
+    def _get_domain(self, change):
```
nit: these two could be made `staticmethod`s

czue commented on corehq/pillows/case_search.py:
```diff
@@ -0,0 +1,151 @@
+
+    def _get_domain(self, change):
+        if isinstance(change, dict):
+            return change.get('key', [])[0]
```
might be worth a comment explaining this. assume has something to do with an assumption about how the data looks during reindexing?

czue commented on corehq/pillows/case_search.py:
```diff
@@ -0,0 +1,151 @@
+        name=pillow_id,
+        document_store=None,
+        checkpoint=checkpoint,
+        change_feed=KafkaChangeFeed(topics=[topics.CASE], group_id='cases-to-es'),
```
I think this should also subscribe to `topics.CASE_SQL` and that might just work out of the box

czue commented on corehq/pillows/case_search.py:
```diff
@@ -0,0 +1,151 @@
+        endkey = [domain, {}, {}]
+        view_kwargs.update({'startkey': startkey, 'endkey': endkey})
+
+    return PillowReindexer(get_case_search_to_elasticsearch_pillow(), CouchViewChangeProvider(
```
in the future we'll need a sql enabled version of this function but if there's not a clear abstraction layer for it now then this is ok to leave for the time being. I think it would just entail checking the domain's sql enabled flag and then using a different change provider though.

proteusvacuum replied:
I don't think this abstraction layer exists yet... I imagine it would be a ChangeProvider that would yield both Couch changes and SQL changes (i.e. smooshing together `CouchViewChangeProvider` and `SqlCaseChangeProvider) so a full reindex would provide all cases. 

It would be great to talk about the best way of going about this in general though. I couldn't figure out a way to pull from couch all the cases that belong to a _list_ of domains (i.e. something like `WHERE domain IN (all_case_search_enabled_domains)`. With the SQL dbaccessors we have, I'm also not sure how we would do that either. (It might not be possible with couch, and might not be worth the effort)

In the instance where the domain is passed in, I think I'd have to make a new type of `SQLCaseChangeProvider` which can take in a dbaccessor method to filter down the cases it yields.

czue replied:
:+1: I think you can always iterate over the domains and then iterate over the cases domain-by-domain, but that's not the nicest.

czue commented on corehq/pillows/case_search.py:
```diff
@@ -0,0 +1,151 @@
+    es = get_es_new()
+    db = CommCareCase.get_db()
+
+    for item in paginate_view(db, 'cases_by_owner/view', 100, startkey=startkey, endkey=endkey, reduce=False):
```
I don't understand why this needs to access couch at all. can't you just run the domain-filtered query in the es index and delete all the results?

czue commented on corehq/pillows/mappings/case_search_mapping.json:
```diff
@@ -0,0 +1,145 @@
+        },
+        "domain": {
+            "fields": {
+                "domain": {
```
when do we ever want fuzzy domain matching?

esoergel replied:
:+1:
We do this in all of the mappings for some reason.

czue commented on corehq/pillows/mappings/case_search_mapping.py:
```diff
@@ -0,0 +1,19 @@
+        mapping = json.loads(data)
+    return mapping
+
+CASE_SEARCH_MAPPING = _CASE_SEARCH_MAPPING()
```
:+1:

esoergel replied:
This means reading the file and performing that substitution will happen whenever `./manage.py X` is run.  Not sure if that's actually a problem or not.  @dannyroberts you've been looking in to start-up optimizations, does this seem like an issue to you?

proteusvacuum replied:
Not sure if it adds too much overhead...
```
In [1]: from corehq.pillows.mappings.case_search_mapping import _CASE_SEARCH_MAPPING

In [2]: %timeit _CASE_SEARCH_MAPPING()
10000 loops, best of 3: 99.2 µs per loop
```

czue commented on urls.py:
```diff
@@ -55,6 +55,7 @@
     (r'^data/', include('corehq.apps.data_interfaces.urls')),
     (r'^', include(hqwebapp_domain_specific)),
     (r'^case/', include('corehq.apps.hqcase.urls')),
+    (r'^case/', include('corehq.apps.case_search.urls')),
```
nit: is there any reason not to give this app it's own base URL? just seems like there's more of a chance for mistakes if pattern matching is happening across two files.

proteusvacuum replied:
No reason, really. I guess I was just trying to keep the urls clean (I think case/search is much nicer than case_search/)

czue commented:
@proteusvacuum this is looking great overall. had a bunch of comments but nothing major. feel free to ping me directly if you want to discuss anything more

benrudolph commented on corehq/apps/case_search/templates/case_search/case_search.html:
```diff
@@ -0,0 +1,186 @@
+{% endblock stylesheets %}
+{% compress js %}
+{% block js %}{{ block.super }}
+    <script src="{% new_static 'codemirror/lib/codemirror.js' %}"></script>
```
nit: we've mapped `new_static` to `static` and are trying to kill `new_static`

proteusvacuum committed:
default dict()

proteusvacuum committed:
rename fuzzy_properties -> get_fuzzy_properties

proteusvacuum committed:
Move js to separate file, add translations

proteusvacuum committed:
Superuser access only to case_search debug view

proteusvacuum committed:
Fix typo: elasticseach -> elasticsearch

proteusvacuum committed:
Fix typo

proteusvacuum committed:
use run instead of change_trigger

proteusvacuum committed:
Add comments

proteusvacuum committed:
Also listen for SQL changes

dimagimon commented on corehq/apps/case_search/tests/test_models.py:
```diff
@@ -0,0 +1,48 @@
+        self.assertItemsEqual(config.get_fuzzy_properties_for_case_type('swashbucklers'), ['has_parrot'])
+
+        config.add_fuzzy_property(case_type="swashbucklers", property="has_sword")
+        self.assertItemsEqual(config.get_fuzzy_properties_for_case_type('swashbucklers'), ['has_parrot', 'has_sword'])
```
E501 line too long (118 > 115 characters)

dimagimon commented on corehq/apps/case_search/static/case_search/js/case_search.js:
```diff
@@ -0,0 +1,55 @@
+                self.results(data.values);
+                var values = JSON.stringify(data.values, null, '    ');
+                if (self.codeMirror === null){
+                    self.codeMirror = CodeMirror( $("#raw-results").get(0), {
```
'CodeMirror' is not defined. (no-undef)

proteusvacuum committed:
lint

proteusvacuum committed:
Use `.dynamic_case_properties`

proteusvacuum committed:
Allow SQL reindexing

Now only fetches pertinent cases from DB

proteusvacuum committed:
Use ES delete_by_query to delete all domain cases

proteusvacuum committed:
Dont allow fuzzy searching domains

fields.exact is needed for use with ESQuery

proteusvacuum committed:
Use JsonObject for CaseSearchConfig

proteusvacuum committed:
Index case properties properly

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into uatbc/case-search-es

dimagimon commented on corehq/pillows/case_search.py:
```diff
@@ -0,0 +1,162 @@
+from corehq.pillows.const import CASE_SEARCH_ALIAS
+from corehq.pillows.mappings.case_search_mapping import CASE_SEARCH_INDEX, \
+    CASE_SEARCH_MAPPING
+from corehq.util.couch_helpers import paginate_view
```
F401 'paginate_view' imported but unused

dimagimon commented on corehq/pillows/case_search.py:
```diff
@@ -0,0 +1,162 @@
+from corehq.apps.change_feed.consumer.feed import KafkaChangeFeed
+from corehq.apps.es import CaseSearchES
+from corehq.elastic import get_es_new
+from corehq.form_processor.models import CommCareCaseSQL
```
F401 'CommCareCaseSQL' imported but unused

proteusvacuum committed:
lint

proteusvacuum commented on corehq/ex-submodules/pillowtop/reindexer/change_providers/case.py:
```diff
@@ -0,0 +1,49 @@
+
+    def iter_changes(self, start_from=None):
+        case_ids = CaseAccessorSQL.get_case_ids_in_domain(self.domain)
+        for case in CaseAccessorSQL.get_cases(case_ids):
```
@snopoke not sure if this is the most efficient way of doing this with SQL cases? When reindexing, I want to fetch all cases only for the pertinent domain.

czue replied:
it's surprising that we don't have a function that returns objects directly but it looks like we don't

snopoke replied:
We don't but would be easy to add - also you should probably chunk this like in `CaseAccessors.iter_cases`

proteusvacuum committed:
Update all-pillow-meta

esoergel commented on corehq/apps/case_search/static/case_search/js/case_search.js:
```diff
@@ -0,0 +1,55 @@
+/* globals hqDefine, ko, $, CodeMirror */
```
:+1:

esoergel commented on corehq/apps/case_search/templates/case_search/case_search.html:
```diff
@@ -0,0 +1,138 @@
+{% block js-inline %}
+    <script>
+     var CaseSearchModel = hqImport('case_search/js/case_search.js');
+     ko.applyBindings(new CaseSearchModel("{% url 'case_details' request.domain '___' %}"), $("#case-search").get(0));
```
We've got a [wrapper](https://github.com/dimagi/commcare-hq/blob/422afba20c2afe19b273b48566b7cf0edd15e58e/corehq/apps/style/static/style/js/hq.helpers.js#L130) for this that is the preferred way to apply bindings.  It'd look like:
```javascript
$("#case-search").koApplyBindings(new CaseSearchModel("{% url 'case_details' request.domain '___' %}"))
```
(Basically it just enforces that you're applying it to precisely one element, which you are here, but we've had problems in the past where that's not the case)

proteusvacuum replied:
:+1:

esoergel commented on corehq/apps/case_search/views.py:
```diff
@@ -0,0 +1,31 @@
+        from corehq.apps.es.case_search import CaseSearchES
+        query = json.loads(request.POST.get('q'))
+        case_type = query.get('type')
+        search_params = query.get('parameters')
```
Might make sense to either provide a fallback `query.get('parameters', [])` or fail hard here if the field is missing, with `query['parameters']`

proteusvacuum replied:
Yeah. This is actually a debug view which actually allows you to search without parameters (i.e. it just searches all cases in the domain).

esoergel replied:
I was thinking that if `parameters` is missing, this line will succeed, and `search_params` will be None, then line 28 will fail on `for param in search_params:`.
If you want to make `parameters` optional, you'd need to default to `[]` or add in an `if search_params:` statement.  If you DO want a hard failure, this line would be the best place for it in my opinion.

proteusvacuum replied:
Gotcha. https://github.com/dimagi/commcare-hq/pull/10981/commits/974918cfe0337bee0b982808f2bf943f8527464c

esoergel commented on corehq/apps/es/case_search.py:
```diff
@@ -0,0 +1,86 @@
+
+
+class CaseSearchES(CaseES):
+    index = CASE_SEARCH_ALIAS
```
:+1:

esoergel commented on corehq/apps/es/es_query.py:
```diff
@@ -250,7 +250,7 @@ def _query(self):
     def set_query(self, query):
         """
-        Add a query.  Most stuff we want is better done with filters, but
+        Set the query.  Most stuff we want is better done with filters, but
```
:+1:

esoergel commented on corehq/apps/es/case_search.py:
```diff
@@ -0,0 +1,86 @@
+                filters.term("{}.key".format(PATH), key)
+            )
+        )
+        return self._add_query(new_query, clause)
```
This looks sufficiently complex that it might be nice to add a test to make sure it outputs the JSON you expect it to output.

proteusvacuum replied:
:+1:

esoergel commented:
This is probably way too late in the game for this kind of talk, but what's the rationale behind a second case index versus extending the primary case index?

proteusvacuum commented:
@esoergel good question -- there were a few reasons. We didn't want to index every property for everyone (indexing with nested properties adds one document per property) as this would dramatically increase the number documents in the case index (not totally sure what the impact of this would be on speed). Since we were only doing this on a domain-by-domain basis and figured that not everyone would use this feature even if / when it was productized, it didn't really make sense to have an index with inconsistent properties (i.e. some cases had all of their case properties indexed, others didn't), so we made a new index.

proteusvacuum committed:
Fix view issues

proteusvacuum committed:
Add test for CaseSearchES

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into uatbc/case-search-es

dimagimon commented on corehq/apps/case_search/static/case_search/js/case_search.js:
```diff
@@ -0,0 +1,55 @@
+                        lineWrapping: true,
+                        viewportMargin: Infinity,
+                        foldGutter: true,
+                        gutters: ["CodeMirror-linenumbers", "CodeMirror-foldgutter"]
```
Missing trailing comma. (comma-dangle)

dimagimon commented on corehq/apps/case_search/static/case_search/js/case_search.js:
```diff
@@ -0,0 +1,55 @@
+            key: "",
+            value: "",
+            clause: "must",
+            fuzzy: false
```
Missing trailing comma. (comma-dangle)

dimagimon commented on corehq/apps/case_search/static/case_search/js/case_search.js:
```diff
@@ -0,0 +1,55 @@
+                key: "",
+                value: "",
+                clause: "must",
+                fuzzy: false
```
Missing trailing comma. (comma-dangle)

proteusvacuum committed:
Comma dangles in js, really? :cake: for dimagimon [skip ci]

proteusvacuum commented:
OK, I think I've resolved all the comments that everyone had here (thanks for the really thorough passes, everyone!). Let me know if you think there is anything else. (Will merge master again when there are new commits to bump travis).

proteusvacuum committed:
Dont fail if there are no parameters

czue commented on corehq/apps/case_search/models.py:
```diff
@@ -19,7 +19,7 @@ class Meta:
     )
     enabled = models.BooleanField(blank=False, null=False, default=False)
-    config = JSONField(default={})
+    config = JSONField(default=dict())
```
I think this is still wrong - should be `dict` the function, not the output of the function

czue commented on corehq/ex-submodules/pillowtop/reindexer/change_providers/case.py:
```diff
@@ -0,0 +1,49 @@
+
+
+class CouchDomainCaseChangeProvider(CouchViewChangeProvider):
+    def __init__(self, domain):
```
is there any reason for this to be a class instead of a function? e.g. `get_couch_case_change_provider(domain)`

proteusvacuum replied:
Yeah, I don't really know what I was thinking here... changed.

czue commented on corehq/pillows/case_search.py:
```diff
@@ -129,27 +127,21 @@ def get_case_search_to_elasticsearch_pillow(pillow_id='CaseSearchToElasticsearch
-    # It initializes the es index if it doesn't exist
+def get_case_search_reindexer(domain=None):
+    """
+    Returns a reindexer that will either all domains with case search enabled, or a single domain
```
something funny about this sentence.

czue commented on corehq/ex-submodules/pillowtop/reindexer/change_providers/case.py:
```diff
@@ -0,0 +1,49 @@
+            yield _sql_case_to_change(case)
+
+
+class DomainCaseChangeProvider(ChangeProvider):
```
non-blocking, but I think this could be generalized to a change provider of change providers and then the details of constructing this particular instance of that could be made a function

czue commented on corehq/pillows/case_search.py:
```diff
@@ -149,11 +150,9 @@ def delete_case_search_cases(domain):
-        if es.exists(CASE_SEARCH_INDEX, CASE_ES_TYPE, item['id']):
-            es.delete(CASE_SEARCH_INDEX, CASE_ES_TYPE, item['id'])
+    query = {'query': CaseSearchES().domain(domain).raw_query['query']}
+    get_es_new().delete_by_query(
```
neat

czue commented on corehq/apps/case_search/models.py:
```diff
@@ -1,7 +1,45 @@
+    properties = jsonobject.ListProperty(unicode)
+
+
+class CaseSearchConfigJSON(jsonobject.JsonObject):
```
nice thanks for changing this. hopefully it feels a bit easier to work with

proteusvacuum replied:
Yeah, that was a great idea.

czue commented:
looking great. we should try and get this in stat (currently merge conflicts)

czue commented on corehq/apps/case_search/models.py:
```diff
@@ -0,0 +1,110 @@
+        primary_key=True
+    )
+    enabled = models.BooleanField(blank=False, null=False, default=False)
+    _config = JSONField(default=dict())
```
commented on the other commit, but this should be `dict` not `dict()`

proteusvacuum committed:
Composite change provider

proteusvacuum committed:
dict() -> dict

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into uatbc/case-search-es

proteusvacuum committed:
Wording changes

czue commented:
build fail looks real (though easy)
```
======================================================================
FAIL: test_prod_config (testapps.test_elasticsearch.tests.test_prod_es_indices.ProdIndexManagementTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/testapps/test_elasticsearch/tests/test_prod_es_indices.py", line 27, in test_prod_config
    self.assertEqual(EXPECTED_PROD_INDICES, found_prod_indices)
AssertionError: Lists differ: [{'alias': 'case_search', 'ind... != [{'alias': u'case_search', 'ty...
First differing element 0:
{'alias': 'case_search', 'index': 'test_case_search_2016-03-15'}
{'alias': u'case_search', 'type': u'case', 'index': u'test_case_search_2016-03-15'}
```

kaapstorm committed:
Add case type to expected index

kaapstorm committed:
Merge branch 'master' into uatbc/case-search-es

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into uatbc/case-search-es

proteusvacuum committed:
Add case_search to ptop_preindex

proteusvacuum committed:
Dont make deploy fail if the table doesn't exist

proteusvacuum committed:
use change object, not couch row

Dependent on https://github.com/dimagi/commcare-hq/pull/11220,
specifically: https://github.com/dimagi/commcare-hq/pull/11220/commits/61b012201f058b867f2fcb090d992b942e47637c

dimagimon commented on corehq/pillows/case_search.py:
```diff
@@ -0,0 +1,158 @@
+from pillowtop.checkpoints.manager import PillowCheckpoint, \
+    PillowCheckpointEventHandler
+from pillowtop.es_utils import ElasticsearchIndexInfo
+from pillowtop.feed.couch import change_from_couch_row
```
F401 'change_from_couch_row' imported but unused

proteusvacuum commented:
(build fail expected until 61b0122 gets merged, but had to push this so that I could get staging working... :gun: :runner: :dart: )

czue commented:
oof, sorry about that.

proteusvacuum commented:
:white_check_mark:
